### PR TITLE
Add support for workers in WASM

### PR DESCRIPTION
**Pull request by oliverdunk**
## Overview

We were panicking when trying to access the locale in a worker. This is because unfortunately, `window` isn't defined, so `Navigator` has to be accessed using `self` instead. This PR implements that to make sure we can reliably use sys-locale in both normal workers as well as service workers.

## Additional thoughts

Ideally I'd like to see this sort of thing solved at a lower level, but in general the ecosystem seems to be somewhat immature in its support for anything other than page contexts. As an example - calling `web_sys::window()` (which returns an option, indicating it will gracefully handle failure) panics in a worker. This makes it hard to gracefully handle and is why my approach here was to implement some new bindings with wasm-bindgen.

I plan to see what we can do to contribute more upstream in the future but I wanted to fix this here to begin with.

## How to test

1. Clone https://github.com/oliverdunk/wasm-worker
2. Checkout the `sys-locale` branch
3. Run `make`
4. Open the server URL printed to the console

In the console, you should see the current locale printed from within a worker.

### CODE REVIEW

**Review by mathieulj — APPROVED** [review]
This looks great, thanks for also updating the docs.

File: src/wasm.rs [hunk 1]
```diff
@@ -1,5 +1,51 @@
...
+        Some(GlobalType::Window(
+            window
+                .dyn_into::<web_sys::Window>()
+                .expect("window was not a Window"),
+        ))
+    } else if let Ok(worker) = Global::get_self() {
+        Some(GlobalType::Worker(
+            worker
+                .dyn_into::<web_sys::WorkerGlobalScope>()
+                .expect("self was not a WorkerGlobalScope"),
```
[the remark this answers is not in the document]
Comment (oliverdunk) [reply src/wasm.rs:39]: You're totally right that we're making some assumptions about things. `self` is actually defined in `window`, but that wouldn't allow us to distinguish between a page and worker, which is why we check `window` first. The reason we need to distinguish between them is that a page has a `Navigator` and a worker has a `WorkerNavigator`, and these are similar but not the same.

I assume your proposal (with a few tweaks to make it real) is something like this:

```rust
pub(crate) fn get() -> Option<String> {
    #[wasm_bindgen]
    extern "C" {
        type Global;

        #[wasm_bindgen(getter, catch, static_method_of = Global, js_class = globalThis, js_name = navigator)]
        fn get_navigator() -> Result<Option<web_sys::Navigator>, JsValue>;
    }

    Global::get_navigator().ok()??.language()
}
```

That does work, but:

- I've had to use `globalThis`, since `global` isn't defined in a page. This unfortunately wouldn't work for us because we currently support browsers that don't implement this.
- We're always casting to `Navigator`, which isn't the real type. It seems like wasm-bindgen lets us get away with this (and it works because both have a language function) but doesn't feel great. If someone copied our code to use for other things on navigator it might not work for them.

We could use `self` instead of `globalThis`, which is defined everywhere, and decide we're comfortable with abusing the Navigator type. I _think_ this would work, although there are definitely caveats to using `self` which is the whole reason `globalThis` was proposed. I don't know for definite if that would be a problem for us (I can't think of any) but it does make me a little reluctant to depend on self given that I know that's not a thing people usually rely on in windows.

So in summary: I definitely share your goal of simplifying things a little, but it's tough.

Side note: For your snippet you moved everything in to `get` which definitely simplifies things. Would we rather do this? I'm happy to do that but it does inline a lot more and make it less reusable in the future.

[the remark this answers is not in the document]
Comment (kjvalencik) [reply src/wasm.rs:39]: * Moving. I only did that for the sake of a simple snippet.
* Excellent point about `globalThis`. I didn't realize we couldn't use that everywhere.
* If think it's safe to cast to the wrong `Navigator` object because we only call the one method; for additional safety, we could forgo the `Navigator` object completely and define our own class with just a single method on it

Overall, I do like the strong typing in your change. My primary suggestion would be tweaking it a bit so that it doesn't panic in non-browser environments.

[the remark this answers is not in the document]
Comment (oliverdunk) [reply src/wasm.rs]: Alright - having discussed this offline a bit, we agreed to keep this approach and actually panic more consistently. The idea is that it's better to throw when you are running the library in an unsupported environment then silently fail and pretend like things are working.

**Review by kjvalencik — APPROVED** [review]

**Comment by kjvalencik** [comment]
Looks great! I like the consistent and clear panic. I think that will make it easier to discover for users in non-browser runtimes.

**Review by complexspaces — APPROVED** [review]
